File: container/lambda_function.py
# lambda_function.py
import asyncio
import boto3
from playwright.async_api import async_playwright

async def scroll_to_bottom(page):
    try:
        previous_height = await page.evaluate('document.body.scrollHeight')
        while True:
            await page.evaluate('window.scrollTo(0, document.body.scrollHeight)')
            await asyncio.sleep(3)
            current_height = await page.evaluate('document.body.scrollHeight')
            if current_height == previous_height:
                break
            previous_height = current_height
    except Exception as e:
        print(f"Error during scrolling: {e}")
        raise

async def download_page_content(url):
    async with async_playwright() as p:
        # Launch the browser with all necessary parameters
        print("Launching browser...")
        browser = await p.chromium.launch(
            headless=True,
            args=[
                "--disable-gpu",
                "--no-sandbox",
                "--single-process",
                "--disable-dev-shm-usage",
                "--no-zygote",
                "--disable-setuid-sandbox",
                "--disable-accelerated-2d-canvas",
                "--disable-dev-shm-usage",
                "--no-first-run",
                "--no-default-browser-check",
                "--disable-background-networking",
                "--disable-background-timer-throttling",
                "--disable-client-side-phishing-detection",
                "--disable-component-update",
                "--disable-default-apps",
                "--disable-domain-reliability",
                "--disable-features=AudioServiceOutOfProcess",
                "--disable-hang-monitor",
                "--disable-ipc-flooding-protection",
                "--disable-popup-blocking",
                "--disable-prompt-on-repost",
                "--disable-renderer-backgrounding",
                "--disable-sync",
                "--force-color-profile=srgb",
                "--metrics-recording-only",
                "--mute-audio",
                "--no-pings",
                "--use-gl=swiftshader",
                "--window-size=1280,1696"
            ]
        )

        context = await browser.new_context(
            user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
        )
        page = await context.new_page()

        # Set headers
        await page.set_extra_http_headers({
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
            "Accept-Language": "en-US,en;q=0.5",
            "Accept-Encoding": "gzip, deflate, br",
            "Connection": "keep-alive",
            "Upgrade-Insecure-Requests": "1",
            "Sec-Fetch-Dest": "document",
            "Sec-Fetch-Mode": "navigate",
            "Sec-Fetch-Site": "none",
            "Sec-Fetch-User": "?1",
            "Cache-Control": "max-age=0",
        })

        await page.set_content("<meta http-equiv='X-Content-Type-Options' content='nosniff'>")

        # Navigate to URL
        print(f"Navigating to URL: {url}")
        try:
            await page.goto(url, timeout=60000)
            try:
                await page.wait_for_load_state('networkidle', timeout=60000)
            except Exception as e:
                print(f"Network idle state timed out: {e}")
            print("Page loaded successfully.")

            # Scroll to bottom
            await scroll_to_bottom(page)
            await page.wait_for_timeout(5000)
            print("Scrolled to the bottom.")

            # Get the page content
            content = await page.content()
        except Exception as e:
            print(f"Failed to load page: {e}")
            raise

        print("Browser closed.")
        return content

async def main(event):
    # Extract parameters from the event payload
    url = event.get('url')
    bucket_name = event.get('bucket')
    output_key = event.get('output_key')
    
    if not url:
        raise ValueError('Error: Missing required parameter (url).')

    # Run the async function to download and process the content
    content = await download_page_content(url)

    # Upload the content to S3
    try:
        print("### Uploading content to S3...")
        print("Content type: ", type(content))
        print("Bucket name: ", bucket_name)
        print("Output key: ", output_key)

        s3_client = boto3.client('s3')
        s3_client.put_object(
            Bucket=bucket_name, 
            Key=output_key, 
            Body=content, 
            ContentType='text/html; charset=utf-8'
        )
    except Exception as e:
        print(f"Error uploading to S3: {e}")
        raise

    print('Source code uploaded successfully')
    return {
        'statusCode': 200,
        'message': 'Source code uploaded successfully',
        'bucket_name': bucket_name,
        'source_code_key': output_key
    }

def handler(event, context):
    return asyncio.run(main(event))

File: container/lambda_function.js
const playwright = require('playwright');
const AWS = require('aws-sdk');
const s3 = new AWS.S3();

async function scrollToBottom(page) {
    try {
        let previousHeight = await page.evaluate(() => document.body.scrollHeight);
        while (true) {
            await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
            await page.waitForTimeout(3000);
            const currentHeight = await page.evaluate(() => document.body.scrollHeight);
            if (currentHeight === previousHeight) {
                break;
            }
            previousHeight = currentHeight;
        }
    } catch (e) {
        console.error(`Error during scrolling: ${e}`);
        throw e;
    }
}

async function downloadPageContent(url) {
    console.log('Starting downloadPageContent...');
    console.log('LAMBDA_TASK_ROOT:', process.env.LAMBDA_TASK_ROOT);
    console.log('Current working directory:', process.cwd());
    
    // 检查目录内容
    const fs = require('fs');
    console.log('Current directory contents:', fs.readdirSync(process.cwd()));
    console.log('/var/task contents:', fs.readdirSync('/var/task'));
    console.log('/var/task/node_modules contents:', fs.readdirSync('/var/task/node_modules'));
    
    // 设置浏览器路径
    const executablePath = '/var/task/node_modules/playwright-core/.local-browsers/chromium-1169/chrome-linux/chrome';
    console.log('Attempting to use browser at:', executablePath);
    
    // 验证浏览器文件是否存在
    try {
        if (fs.existsSync(executablePath)) {
            console.log('Chrome executable exists');
            console.log('File permissions:', fs.statSync(executablePath).mode.toString(8));
        } else {
            console.log('Chrome executable not found');
            console.log('Checking alternative locations...');
            console.log('/root/.cache/ms-playwright contents:', fs.readdirSync('/root/.cache/ms-playwright'));
        }
    } catch (e) {
        console.log('Error checking chrome executable:', e);
    }
    
    console.log('Launching browser...');
    const browser = await playwright.chromium.launch({
        headless: true,
        executablePath,
        args: [
            '--disable-gpu',
            '--no-sandbox',
            '--single-process',
            '--disable-dev-shm-usage',
            '--no-zygote',
            '--disable-setuid-sandbox',
            '--disable-accelerated-2d-canvas',
            '--no-first-run',
            '--no-default-browser-check',
            '--disable-background-networking',
            '--disable-background-timer-throttling',
            '--disable-client-side-phishing-detection',
            '--disable-component-update',
            '--disable-default-apps',
            '--disable-domain-reliability',
            '--disable-features=AudioServiceOutOfProcess',
            '--disable-hang-monitor',
            '--disable-ipc-flooding-protection',
            '--disable-popup-blocking',
            '--disable-prompt-on-repost',
            '--disable-renderer-backgrounding',
            '--disable-sync',
            '--force-color-profile=srgb',
            '--metrics-recording-only',
            '--mute-audio',
            '--no-pings',
            '--use-gl=swiftshader',
            '--window-size=1280,1696'
        ]
    });

    const context = await browser.newContext({
        userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    });

    const page = await context.newPage();

    // Set headers
    await page.setExtraHTTPHeaders({
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.5',
        'Accept-Encoding': 'gzip, deflate, br',
        'Connection': 'keep-alive',
        'Upgrade-Insecure-Requests': '1',
        'Sec-Fetch-Dest': 'document',
        'Sec-Fetch-Mode': 'navigate',
        'Sec-Fetch-Site': 'none',
        'Sec-Fetch-User': '?1',
        'Cache-Control': 'max-age=0'
    });

    try {
        console.log(`Navigating to URL: ${url}`);
        await page.goto(url, { timeout: 60000 });
        
        try {
            await page.waitForLoadState('networkidle', { timeout: 60000 });
        } catch (e) {
            console.log(`Network idle state timed out: ${e}`);
        }
        
        console.log('Page loaded successfully.');

        // Scroll to bottom
        await scrollToBottom(page);
        await page.waitForTimeout(5000);
        console.log('Scrolled to the bottom.');

        // Take screenshot
        console.log('Taking screenshot...');
        const screenshot = await page.screenshot({ fullPage: true });
        console.log('Screenshot captured.');

        // Get page content
        const content = await page.content();

        await browser.close();
        console.log('Browser closed.');

        return {
            content,
            screenshot
        };
    } catch (e) {
        console.error(`Failed to load page: ${e}`);
        await browser.close();
        throw e;
    }
}

async function uploadToS3(bucket, key, body, contentType) {
    const params = {
        Bucket: bucket,
        Key: key,
        Body: body,
        ContentType: contentType
    };

    try {
        await s3.putObject(params).promise();
        console.log(`Successfully uploaded to S3: ${key}`);
    } catch (e) {
        console.error(`Error uploading to S3: ${e}`);
        throw e;
    }
}

// 本地测试函数
async function localTest() {
    const testEvent = {
        url: 'https://www.example.com',
        bucket: 'test-bucket',
        outputKey: 'test-output'
    };
    
    try {
        const result = await exports.handler(testEvent);
        console.log('Test result:', result);
    } catch (e) {
        console.error('Test failed:', e);
    }
}

exports.handler = async (event) => {
    const url = event.url;
    const bucket = event.bucket;
    const outputKey = event.outputKey;

    if (!url) {
        throw new Error('Missing required parameter (url)');
    }

    try {
        const { content, screenshot } = await downloadPageContent(url);

        // 上传HTML内容
        await uploadToS3(
            bucket,
            `${outputKey}.html`,
            content,
            'text/html; charset=utf-8'
        );

        // 上传截图
        await uploadToS3(
            bucket,
            `${outputKey}.png`,
            screenshot,
            'image/png'
        );

        return {
            statusCode: 200,
            message: 'Content and screenshot uploaded successfully',
            bucket: bucket,
            htmlKey: `${outputKey}.html`,
            screenshotKey: `${outputKey}.png`
        };
    } catch (error) {
        console.error('Error:', error);
        throw error;
    }
};

// 如果直接运行文件，执行本地测试
if (require.main === module) {
    localTest();
}
